In [2]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

df = pd.read_csv("../data/online_retail_ii.csv")

df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [3]:
df.columns
df.info()
df.describe(include="all")

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 136.6 MB


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
count,1067371,1067371,1062989,1.067371e+06,1067371,1.067371e+06,824364.000000,1067371
unique,53628,5305,5698,NaN,47635,NaN,NaN,43
top,537434,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,2010-12-06 16:57:00,NaN,NaN,United Kingdom
freq,1350,5829,5918,NaN,1350,NaN,NaN,981330
mean,NaN,NaN,NaN,9.938898e+00,NaN,4.649388e+00,15324.638504,NaN
std,NaN,NaN,NaN,1.727058e+02,NaN,1.235531e+02,1697.464450,NaN
min,NaN,NaN,NaN,-8.099500e+04,NaN,-5.359436e+04,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000e+00,NaN,1.250000e+00,13975.000000,NaN
50%,NaN,NaN,NaN,3.000000e+00,NaN,2.100000e+00,15255.000000,NaN
75%,NaN,NaN,NaN,1.000000e+01,NaN,4.150000e+00,16797.000000,NaN


In [4]:
df.isna().sum().sort_values(ascending=False)
(df.isna().mean()*100).round(2)

Invoice         0.00
StockCode       0.00
Description     0.41
Quantity        0.00
InvoiceDate     0.00
Price           0.00
Customer ID    22.77
Country         0.00
dtype: float64

custom id 缺失率高

In [5]:
df.duplicated().sum()
df[df.duplicated()].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


In [6]:
df["Invoice"].value_counts().head(30)


Invoice
537434    1350
538071    1304
537638    1202
537237    1194
536876    1186
          ... 
576617     593
576837     585
536031     582
490074     580
491966     579
Name: count, Length: 30, dtype: int64

In [7]:
df["Invoice"].astype(str).str[1].value_counts()

Invoice
3    140751
7    131743
5    123122
6    117434
4    113310
2    108331
9    101824
1     99831
0     99626
8     31399
Name: count, dtype: int64

In [8]:
df["Invoice"].astype(str).str[0].value_counts()

Invoice
5    939382
4    108489
C     19494
A         6
Name: count, dtype: int64

 # 为什么前缀是CA

In [9]:
df["StockCode"].value_counts(50)
df["StockCode"]

0           85048
1          79323P
2          79323W
3           22041
4           21232
            ...  
1067366     22899
1067367     23254
1067368     23255
1067369     22138
1067370      POST
Name: StockCode, Length: 1067371, dtype: str

In [10]:
df["Quantity"].describe()
df[df["Quantity"]<0]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
...,...,...,...,...,...,...,...,...
1065910,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
1067002,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
1067176,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
1067177,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


In [11]:
df["Price"].describe()

count    1.067371e+06
mean     4.649388e+00
std      1.235531e+02
min     -5.359436e+04
25%      1.250000e+00
50%      2.100000e+00
75%      4.150000e+00
max      3.897000e+04
Name: Price, dtype: float64

In [12]:
df["Customer ID"].isna().mean()
df[df["Customer ID"].isna()]["Country"].value_counts()
df["Country"].value_counts()

Country
United Kingdom    981330
EIRE               17866
Germany            17624
France             14330
Netherlands         5140
                   ...  
West Indies           54
Bermuda               34
Nigeria               32
Czech Republic        30
Saudi Arabia          10
Name: count, Length: 43, dtype: int64

In [13]:
df[df["Invoice"].astype(str).str.startswith("C")][
    ["Invoice","Description","Quantity"]
]

,Invoice,Description,Quantity
178,C489449,PAPER BUNTING WHITE LACE,-12
179,C489449,CREAM FELT EASTER EGG BASKET,-6
180,C489449,POTTING SHED SOW 'N' GROW SET,-4
181,C489449,POTTING SHED TWINE,-6
182,C489449,PAPER CHAIN KIT RETRO SPOT,-12
...,...,...,...
1065910,C581490,ZINC T-LIGHT HOLDER STARS SMALL,-11
1067002,C581499,Manual,-1
1067176,C581568,VICTORIAN SEWING BOX LARGE,-5
1067177,C581569,HANGING HEART JAR T-LIGHT HOLDER,-1


这里全是负数，有一个很明显的outlier

官网原文：InvoiceNo: Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.
StockCode: Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.
InvoiceDate: Invice date and time. Numeric. The day and time when a transaction was generated.
UnitPrice: Unit price. Numeric. Product price per unit in sterling (£).
CustomerID: Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal. The name of the country where a customer resides.

In [14]:
df[df["Invoice"].astype(str).str.startswith("A")][
    ["Invoice","Description","Quantity"]
]

,Invoice,Description,Quantity
179403,A506401,Adjust bad debt,1
276274,A516228,Adjust bad debt,1
403472,A528059,Adjust bad debt,1
825443,A563185,Adjust bad debt,1
825444,A563186,Adjust bad debt,1
825445,A563187,Adjust bad debt,1


In [15]:
# 统一派生字段：后续所有统计都基于这个副本，不改动原始 df。
work = df.assign(
    InvoiceStr=df["Invoice"].astype(str),
    StockCodeStr=df["StockCode"].astype(str).str.strip(),
    InvoiceDateParsed=pd.to_datetime(df["InvoiceDate"]),
    LineAmount=df["Quantity"] * df["Price"],
)
work = work.assign(
    InvoicePrefix=work["InvoiceStr"].str.extract(r"^([A-Za-z]+)", expand=False).fillna("numeric"),
    IsCancellation=work["InvoiceStr"].str.startswith("C", na=False),
    IsBadDebtAdjustment=work["InvoiceStr"].str.startswith("A", na=False),
    IsCustomerMissing=work["Customer ID"].isna(),
    IsDescriptionMissing=work["Description"].isna(),
)
work = work.assign(
    IsDamageOrInventoryLoss=(
        work["Price"].eq(0)
        & work["Quantity"].lt(0)
        & ~work["IsCancellation"]
        & ~work["IsBadDebtAdjustment"]
    )
)

def pct(value, denominator):
    return round(value / denominator * 100, 2) if denominator else np.nan

total_rows = len(work)
total_amount = work["LineAmount"].sum()
print(f"rows={total_rows:,}, net line amount={total_amount:,.2f}")


rows=1,067,371, net line amount=19,287,250.57


### 1. Invoice 前缀是交易类型

`C` 是取消/退货；`A` 是 bad debt adjustment。这里需要先定义销售口径：

- **Gross product sales**：通常只统计正向商品销售，排除 `C` 取消单、`A` bad debt、费用/邮费/折扣等非商品行。
- **Net product sales**：可包含 `C` 取消/退货来抵减商品销售，但仍应把 `A` bad debt 和费用类 StockCode 单独列示。
- **财务总额**：可以保留费用、邮费、佣金、bad debt，但要单独分组，否则会污染商品销量和客单价分析。

本数据中 `C` 有 **19,494 行**，金额合计 **-1,526,667.86**；`A` 只有 **6 行**，都是 `Adjust bad debt`，其中 **5 行 Price 为负数**。

In [16]:
invoice_prefix_summary = (
    work.groupby("InvoicePrefix", dropna=False)
    .agg(
        rows=("Invoice", "size"),
        quantity_sum=("Quantity", "sum"),
        amount_sum=("LineAmount", "sum"),
        negative_price_rows=("Price", lambda s: (s < 0).sum()),
        zero_price_rows=("Price", lambda s: (s == 0).sum()),
        missing_customer_rows=("Customer ID", lambda s: s.isna().sum()),
    )
    .sort_values("rows", ascending=False)
)
display(invoice_prefix_summary)

display(
    work.loc[
        work["IsBadDebtAdjustment"],
        ["Invoice", "StockCode", "Description", "Quantity", "InvoiceDate", "Price", "Customer ID", "Country", "LineAmount"],
    ]
)

,rows,quantity_sum,amount_sum,negative_price_rows,zero_price_rows,missing_customer_rows
InvoicePrefix,,,,,,
numeric,1047871,11099478,2.096153e+07,0,6202,242251
C,19494,-490992,-1.526668e+06,0,0,750
A,6,6,-1.476141e+05,5,0,6


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,LineAmount
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom,-53594.36
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom,-44031.79
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom,-38925.87
825443,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom,11062.06
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom,-11062.06
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom,-11062.06


### 2. 非产品 StockCode 需要分组，而不是简单 value_counts

特殊 StockCode 可以解释为邮费/运输、折扣、手工调整、平台费用/佣金、礼品券、样品、测试或内部 dotcom 代码。按“不是数字开头”的宽口径统计，本数据有 **6,093 行**（约 **0.57%**），金额合计 **-95,225.44**，占净金额约 **-0.49%**。这说明它们行数占比小，但会直接影响金额口径。

如果只统计题目中最明确的 `POST / DOT / C2 / D / M / BANK CHARGES / AMAZONFEE / CRUK / gift_0001_*`，共有 **5,709 行**；差异主要来自 `S` samples、`ADJUST/ADJUST2`、`B` bad debt、`TEST001/TEST002`、`PADS` 和 `DCGS*` 内部/dotcom 代码。

In [17]:
desc_count = (
    work.groupby("StockCodeStr")["Description"]
    .nunique()
    .sort_values(ascending=False)
)

desc_count


StockCodeStr
20713      9
22423      7
23084      7
21181      7
22734      7
          ..
85148      0
71263G     0
71265      0
71406B     0
72024HC    0
Name: Description, Length: 5304, dtype: int64

In [18]:
code = desc_count[desc_count == 0].index[0]

work.loc[
    work["StockCodeStr"] == code,
    ["StockCodeStr", "Description"]
]

,StockCodeStr,Description
65203,35813A,NaN


In [19]:
missing_desc = (
    work.groupby("StockCodeStr")["Description"]
    .apply(lambda s: s.isna().all())
)

missing_desc[missing_desc]

StockCodeStr
10124C          True
15060A          True
16065A          True
16131           True
16132           True
                ... 
DCGSLBOY        True
DCGSLGIRL       True
GIFT            True
gift_0001_60    True
gift_0001_90    True
Name: Description, Length: 355, dtype: bool

In [20]:
work.loc[
    work["Description"].isna(),
    ["StockCodeStr", "InvoiceStr", "Quantity", "Price"]
]

,StockCodeStr,InvoiceStr,Quantity,Price
470,21646,489521,-50,0.0
3114,20683,489655,-44,0.0
3161,21350,489659,230,0.0
3731,84292,489781,17,0.0
4296,18010,489806,-770,0.0
...,...,...,...,...
1060783,84581,581199,-2,0.0
1060787,23406,581203,15,0.0
1060793,21620,581209,6,0.0
1062442,72817,581234,27,0.0


In [34]:
work.loc[work["IsDamageOrInventoryLoss"]].shape

(3457, 21)

In [22]:
work.loc[work["IsDamageOrInventoryLoss"],
    ["InvoiceStr", "StockCodeStr", "Description", "Quantity", "Price"]
].head(30)

,InvoiceStr,StockCodeStr,Description,Quantity,Price
263,489464,21733,85123a mixed,-96,0.0
283,489463,71477,short,-240,0.0
284,489467,85123A,21733 mixed,-192,0.0
470,489521,21646,NaN,-50,0.0
3114,489655,20683,NaN,-44,0.0
...,...,...,...,...,...
16185,490735,35976B,NaN,-60,0.0
16186,490736,35983,NaN,-200,0.0
17427,490765,21450,damaged,-31,0.0
17428,490766,51008,wet,-200,0.0


In [23]:
(
    work.loc[work["IsDamageOrInventoryLoss"], "IsCancellation"]
    .value_counts()
)

IsCancellation
False    3457
Name: count, dtype: int64

In [24]:
work.loc[work["IsDamageOrInventoryLoss"],
    [
        "InvoiceStr",
        "StockCodeStr",
        "Description",
        "Quantity",
        "Customer ID",
    ]
].head(30)

,InvoiceStr,StockCodeStr,Description,Quantity,Customer ID
263,489464,21733,85123a mixed,-96,NaN
283,489463,71477,short,-240,NaN
284,489467,85123A,21733 mixed,-192,NaN
470,489521,21646,NaN,-50,NaN
3114,489655,20683,NaN,-44,NaN
...,...,...,...,...,...
16185,490735,35976B,NaN,-60,NaN
16186,490736,35983,NaN,-200,NaN
17427,490765,21450,damaged,-31,NaN
17428,490766,51008,wet,-200,NaN


In [25]:
(
    work.loc[work["IsDamageOrInventoryLoss"], "Description"]
        .fillna("<missing>")
        .value_counts()
        .head(30)
)

Description
<missing>     2689
check          123
damages         84
?               83
damaged         78
              ... 
Dotcom           3
broken           3
CHECK            3
wet pallet       3
lost             2
Name: count, Length: 30, dtype: int64

In [26]:
import re


def classify_stock_code(code_value):
    code_text = str(code_value).strip().upper()
    if code_text in {"POST", "DOT", "C2"}:
        return "postage_carriage_dotcom"
    if code_text == "D":
        return "discount"
    if code_text in {"M", "ADJUST", "ADJUST2"}:
        return "manual_adjustment"
    if code_text in {"BANK CHARGES", "AMAZONFEE", "CRUK", "B"}:
        return "fees_commission_bad_debt"
    if code_text.startswith("GIFT"):
        return "gift_voucher"
    if code_text in {"S", "TEST001", "TEST002", "PADS"} or code_text.startswith("DCGS"):
        return "sample_test_internal"
    if re.fullmatch(r"\d+[A-Z]*", code_text):
        return "product_like"
    return "other_non_product_or_unknown"

work = work.assign(StockCodeCategory=work["StockCodeStr"].map(classify_stock_code))

stock_category_summary = (
    work.groupby("StockCodeCategory")
    .agg(
        rows=("StockCode", "size"),
        distinct_codes=("StockCodeStr", "nunique"),
        amount_sum=("LineAmount", "sum"),
    )
    .assign(
        row_pct=lambda x: (x["rows"] / total_rows * 100).round(3),
        amount_pct=lambda x: (x["amount_sum"] / total_amount * 100).round(3),
    )
    .sort_values("rows", ascending=False)
)
display(stock_category_summary)

non_product_code_list = (
    work.loc[~work["StockCodeCategory"].eq("product_like")]
    .groupby("StockCodeCategory")["StockCodeStr"]
    .agg(lambda s: ", ".join(sorted(s.astype(str).unique())))
    .reset_index(name="codes")
)
display(non_product_code_list)

work = work.assign(
    IsGrossProductSale=(
        work["InvoicePrefix"].eq("numeric")
        & work["StockCodeCategory"].eq("product_like")
        & work["Quantity"].gt(0)
        & work["Price"].gt(0)
    ),
    IsNetProductLine=(
        work["StockCodeCategory"].eq("product_like")
        & ~work["IsBadDebtAdjustment"]
        & work["Price"].gt(0)
        & work["Quantity"].ne(0)
    ),
)

sales_measure_summary = pd.DataFrame(
    [
        {"measure": "gross_product_sales", "amount": work.loc[work["IsGrossProductSale"], "LineAmount"].sum()},
        {"measure": "net_product_sales", "amount": work.loc[work["IsNetProductLine"], "LineAmount"].sum()},
        {"measure": "financial_total", "amount": work["LineAmount"].sum()},
    ]
)
display(sales_measure_summary)


,rows,distinct_codes,amount_sum,row_pct,amount_pct
StockCodeCategory,,,,,
product_like,1061278,5242,1.938248e+07,99.429,100.494
postage_carriage_dotcom,3850,3,4.483745e+05,0.361,2.325
manual_adjustment,1496,4,-7.521498e+04,0.140,-0.390
sample_test_internal,298,38,-4.728033e+03,0.028,-0.025
discount,177,1,-1.348454e+04,0.017,-0.070
fees_commission_bad_debt,167,4,-4.518737e+05,0.016,-2.343
gift_voucher,101,10,1.686610e+03,0.009,0.009
other_non_product_or_unknown,4,2,1.475000e+01,0.000,0.000


,StockCodeCategory,codes
0,discount,D
1,fees_commission_bad_debt,"AMAZONFEE, B, BANK CHARGES, CRUK"
2,gift_voucher,"GIFT, gift_0001_10, gift_0001_20, gift_0001_30..."
3,manual_adjustment,"ADJUST, ADJUST2, M, m"
4,other_non_product_or_unknown,"C3, SP1002"
5,postage_carriage_dotcom,"C2, DOT, POST"
6,sample_test_internal,"DCGS0003, DCGS0004, DCGS0006, DCGS0016, DCGS00..."


,measure,amount
0,gross_product_sales,2.010900e+07
1,net_product_sales,1.938248e+07
2,financial_total,1.928725e+07


### 3. 极端 Quantity 查询

最大 `Quantity=80,995` 和最小 `Quantity=-80,995` 是同一商品、同一客户的销售与取消，间隔只有 **12 分钟**：

- `581483 / 23843 / 80995 / Customer 16446 / 2011-12-09 09:15`
- `C581484 / 23843 / -80995 / Customer 16446 / 2011-12-09 09:27`

这是一条很好的数据质量案例：单看箱线图它是极端异常；结合 Invoice 前缀和时间，它更像一笔大额误下单后被快速取消。处理策略应该是**配对解释和保留审计痕迹**，而不是无脑删除。

In [27]:
extreme_quantity_rows = work.loc[
    work["Quantity"].isin([work["Quantity"].max(), work["Quantity"].min()]),
    ["Invoice", "StockCode", "Description", "Quantity", "InvoiceDateParsed", "Price", "Customer ID", "Country", "LineAmount"],
].sort_values("InvoiceDateParsed")

display(extreme_quantity_rows)

if len(extreme_quantity_rows) == 2:
    minutes_between = extreme_quantity_rows["InvoiceDateParsed"].diff().dt.total_seconds().iloc[-1] / 60
    print(f"minutes_between={minutes_between:.0f}")

,Invoice,StockCode,Description,Quantity,InvoiceDateParsed,Price,Customer ID,Country,LineAmount
1065882,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom,168469.6
1065883,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2011-12-09 09:27:00,2.08,16446.0,United Kingdom,-168469.6


minutes_between=12


### 4. 负价格、零价格、负数量要分开解释

价格异常不是一种问题，其中 `Price = 0` 且 `Quantity < 0` 的数字 Invoice 需要单独解释为商品损坏/库存损耗，而不是客户退货：

- `Price < 0` 只有 **5 行**，全部是 `A` 开头的 bad debt adjustment，不是商品负价。
- `Price = 0` 有 **6,202 行**，大量描述是 `lost`、`damaged`、`damages`、`wet`、`short`、`check` 或缺失描述，更像商品损坏、库存损耗、盘点差异和人工检查。
- `Quantity < 0` 不全是取消单：有 **3,457 行**是数字 Invoice、`Price = 0`、负数量，已标记为 `IsDamageOrInventoryLoss`。这些记录基本不影响销售金额，但会严重影响库存/销量分析。

因此清洗时不能只写 `Quantity < 0` 或 `Price <= 0` 删除，而要按交易语义分桶：客户退货走取消/退货口径；这批零价负数量记录走损坏/库存损耗口径。

In [ ]:
price_quantity_quality = pd.DataFrame(
    [
        {"issue": "Price < 0", "rows": int((work["Price"] < 0).sum())},
        {"issue": "Price = 0", "rows": int((work["Price"] == 0).sum())},
        {"issue": "Damage / inventory loss: Price = 0 and Quantity < 0", "rows": int(work["IsDamageOrInventoryLoss"].sum())},
    ]
)
display(price_quantity_quality)

display(
    work.loc[work["Price"] < 0, ["Invoice", "StockCode", "Description", "Quantity", "InvoiceDate", "Price", "Customer ID", "LineAmount"]]
)

display(
    work.loc[work["IsDamageOrInventoryLoss"], "Description"]
    .fillna("<missing>")
    .str.upper()
    .value_counts()
    .head(20)
    .rename_axis("Description")
    .reset_index(name="rows")
)

In [ ]:
customer_missing_summary = pd.DataFrame(
    [
        {
            "segment": "Customer ID missing",
            "rows": int(work["IsCustomerMissing"].sum()),
            "row_pct": pct(work["IsCustomerMissing"].sum(), total_rows),
            "amount_sum": work.loc[work["IsCustomerMissing"], "LineAmount"].sum(),
            "amount_pct_of_net": pct(work.loc[work["IsCustomerMissing"], "LineAmount"].sum(), total_amount),
            "uk_pct": pct(work.loc[work["IsCustomerMissing"], "Country"].eq("United Kingdom").sum(), work["IsCustomerMissing"].sum()),
            "avg_line_amount": work.loc[work["IsCustomerMissing"], "LineAmount"].mean(),
        },
        {
            "segment": "Customer ID present",
            "rows": int((~work["IsCustomerMissing"]).sum()),
            "row_pct": pct((~work["IsCustomerMissing"]).sum(), total_rows),
            "amount_sum": work.loc[~work["IsCustomerMissing"], "LineAmount"].sum(),
            "amount_pct_of_net": pct(work.loc[~work["IsCustomerMissing"], "LineAmount"].sum(), total_amount),
            "uk_pct": pct(work.loc[~work["IsCustomerMissing"], "Country"].eq("United Kingdom").sum(), (~work["IsCustomerMissing"]).sum()),
            "avg_line_amount": work.loc[~work["IsCustomerMissing"], "LineAmount"].mean(),
        },
    ]
)
display(customer_missing_summary)

display(
    work.loc[work["IsCustomerMissing"]]
    .groupby(["StockCodeStr", "Description"], dropna=False)
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"))
    .sort_values("rows", ascending=False)
    .head(20)
)

### 5b. Customer ID 缺失群体深挖

进一步检查后，`Customer ID` 缺失不能简单等同于商品损坏/库存损耗。更准确的结论是：**商品损坏/库存损耗是缺失 Customer ID 群体里的高置信子集，但缺失群体的主体仍是正常正向交易**。

关键证据：

- 缺失发生在 **invoice 级别**：没有一张 invoice 同时混合“有 Customer ID”和“缺 Customer ID”，说明这更像订单/渠道层面的客户绑定缺失，而不是单行字段随机坏掉。
- `Price = 0 & Quantity < 0` 的商品损坏/库存损耗共有 **3,457 行**，这些行 **100% 缺失 Customer ID**。
- 但这 3,457 行只占全部缺失 Customer ID 记录的 **1.42%**；缺失群体里约 **97.17%** 仍是 `Quantity > 0 & Price > 0` 的正常正向销售。
- 缺失群体 **98.77%** 来自 UK，并且 `DOTCOM POSTAGE` / `DOT` 高度集中在缺失群体，说明它可能包含匿名线上订单、dotcom/system channel、未绑定客户订单，以及库存损耗/系统调整。

因此，销售报表不能直接删除缺失 Customer ID 的记录；RFM 客户分析应排除或单独归入 `anonymous/system channel`；库存质量分析则应保留并单独追踪其中的损坏/库存损耗子集。


In [30]:
# Customer ID 缺失深挖：它是订单/渠道层面的缺失，不是单纯商品损耗。
invoice_customer_completeness = (
    work.groupby("InvoiceStr")
    .agg(
        rows=("Invoice", "size"),
        missing_customer_rows=("IsCustomerMissing", "sum"),
        amount_sum=("LineAmount", "sum"),
        has_dotcom_postage=("StockCodeStr", lambda s: s.astype(str).str.upper().eq("DOT").any()),
        damage_loss_rows=("IsDamageOrInventoryLoss", "sum"),
    )
    .reset_index()
)
invoice_customer_completeness = invoice_customer_completeness.assign(
    customer_id_pattern=np.select(
        [
            invoice_customer_completeness["missing_customer_rows"].eq(0),
            invoice_customer_completeness["missing_customer_rows"].eq(invoice_customer_completeness["rows"]),
        ],
        ["all_present", "all_missing"],
        default="mixed_missing_present",
    )
)

display(
    invoice_customer_completeness.groupby("customer_id_pattern")
    .agg(
        invoices=("InvoiceStr", "size"),
        rows=("rows", "sum"),
        amount_sum=("amount_sum", "sum"),
        dotcom_postage_invoices=("has_dotcom_postage", "sum"),
        damage_loss_invoices=("damage_loss_rows", lambda s: int((s > 0).sum())),
    )
    .assign(
        invoice_pct=lambda x: (x["invoices"] / x["invoices"].sum() * 100).round(2),
        row_pct=lambda x: (x["rows"] / total_rows * 100).round(2),
    )
)

missing_customer = work.loc[work["IsCustomerMissing"]]
customer_missing_buckets = pd.DataFrame(
    [
        {
            "bucket": "normal positive line: Quantity > 0 and Price > 0",
            "rows": int((missing_customer["Quantity"].gt(0) & missing_customer["Price"].gt(0)).sum()),
            "amount_sum": missing_customer.loc[
                missing_customer["Quantity"].gt(0) & missing_customer["Price"].gt(0), "LineAmount"
            ].sum(),
        },
        {
            "bucket": "damage / inventory loss: Price = 0 and Quantity < 0",
            "rows": int(missing_customer["IsDamageOrInventoryLoss"].sum()),
            "amount_sum": missing_customer.loc[missing_customer["IsDamageOrInventoryLoss"], "LineAmount"].sum(),
        },
        {
            "bucket": "zero-price positive quantity",
            "rows": int((missing_customer["Price"].eq(0) & missing_customer["Quantity"].gt(0)).sum()),
            "amount_sum": 0.0,
        },
        {
            "bucket": "cancellation / return: C invoice",
            "rows": int(missing_customer["IsCancellation"].sum()),
            "amount_sum": missing_customer.loc[missing_customer["IsCancellation"], "LineAmount"].sum(),
        },
        {
            "bucket": "bad debt: A invoice",
            "rows": int(missing_customer["IsBadDebtAdjustment"].sum()),
            "amount_sum": missing_customer.loc[missing_customer["IsBadDebtAdjustment"], "LineAmount"].sum(),
        },
    ]
)
customer_missing_buckets = customer_missing_buckets.assign(
    pct_of_missing_rows=(customer_missing_buckets["rows"] / len(missing_customer) * 100).round(2),
    amount_sum=customer_missing_buckets["amount_sum"].round(2),
)
display(customer_missing_buckets)

display(
    missing_customer.groupby(["StockCodeStr", "Description"], dropna=False)
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"), quantity_sum=("Quantity", "sum"))
    .sort_values("rows", ascending=False)
    .head(12)
    .reset_index()
)

print(
    "Damage/loss rows with missing Customer ID:",
    int((work["IsDamageOrInventoryLoss"] & work["IsCustomerMissing"]).sum()),
    "/",
    int(work["IsDamageOrInventoryLoss"].sum()),
)
print(
    "Damage/loss share within missing Customer ID rows:",
    pct((work["IsDamageOrInventoryLoss"] & work["IsCustomerMissing"]).sum(), work["IsCustomerMissing"].sum()),
    "%",
)
print(
    "Description missing rows with missing Customer ID:",
    int((work["IsDescriptionMissing"] & work["IsCustomerMissing"]).sum()),
    "/",
    int(work["IsDescriptionMissing"].sum()),
)



,invoices,rows,amount_sum,dotcom_postage_invoices,damage_loss_invoices,invoice_pct,row_pct
customer_id_pattern,,,,,,,
all_missing,8752,243007,2.638958e+06,1409,3393,16.32,22.77
all_present,44876,824364,1.664829e+07,16,0,83.68,77.23


,bucket,rows,amount_sum,pct_of_missing_rows
0,normal positive line: Quantity > 0 and Price > 0,236122,3229538.96,97.17
1,damage / inventory loss: Price = 0 and Quantit...,3457,0.00,1.42
2,zero-price positive quantity,2674,0.00,1.10
3,cancellation / return: C invoice,750,-431531.07,0.31
4,bad debt: A invoice,6,-147614.08,0.00


,StockCodeStr,Description,rows,amount_sum,quantity_sum
0,DOT,DOTCOM POSTAGE,1428,310741.11,1422
1,85099B,JUMBO BAG RED RETROSPOT,668,11843.84,3107
2,22423,REGENCY CAKESTAND 3 TIER,635,58076.95,2680
3,21931,JUMBO STORAGE BAG SUKI,635,14029.48,3266
4,47566,PARTY BUNTING,621,45262.27,4775
5,22379,RECYCLING BAG RETROSPOT,573,7031.88,1568
6,22411,JUMBO SHOPPER VINTAGE RED PAISLEY,529,12467.42,3031
7,20719,WOODLAND CHARLOTTE BAG,526,13051.73,5660
8,20712,JUMBO BAG WOODLAND ANIMALS,509,8525.39,2028
9,21232,STRAWBERRY CERAMIC TRINKET BOX,507,7951.38,2886


Damage/loss rows with missing Customer ID: 3457 / 3457
Damage/loss share within missing Customer ID rows: 1.42 %
Description missing rows with missing Customer ID: 4382 / 4382


In [35]:
positive_missing = missing_customer.loc[
    missing_customer["Quantity"].gt(0) & missing_customer["Price"].gt(0)
].copy()

# ===== 1. 国家分布 =====
display(
    positive_missing.groupby("Country")
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"), invoices=("InvoiceStr", "nunique"))
    .assign(
        row_pct=lambda x: (x["rows"] / len(positive_missing) * 100).round(2),
        amount_pct=lambda x: (x["amount_sum"] / positive_missing["LineAmount"].sum() * 100).round(2),
    )
    .sort_values("rows", ascending=False)
)

# ===== 2. 发票层面：这些行属于什么模式的发票？ =====
positive_missing_invoices = positive_missing["InvoiceStr"].unique()
invoice_patterns = invoice_customer_completeness.loc[
    invoice_customer_completeness["InvoiceStr"].isin(positive_missing_invoices)
]
display(
    invoice_patterns["customer_id_pattern"].value_counts().to_frame("invoices").assign(
        pct=lambda x: (x["invoices"] / x["invoices"].sum() * 100).round(2)
    )
)

# ===== 3. 整单缺失 vs 部分缺失，分别是什么特征 =====
positive_missing = positive_missing.assign(
    invoice_pattern=positive_missing["InvoiceStr"].map(
        dict(zip(invoice_customer_completeness["InvoiceStr"], invoice_customer_completeness["customer_id_pattern"]))
    )
)
# 整单缺失的发票含运费情况
all_missing_invoices = invoice_customer_completeness.loc[
    invoice_customer_completeness["customer_id_pattern"].eq("all_missing")
]
print("整单缺失中运单(DOT)占比:",
      pct(all_missing_invoices["has_dotcom_postage"].sum(), len(all_missing_invoices)), "%")
# 混合缺失的发票——同一张发票里有的行有Customer有的没有
mixed_invoices = positive_missing.loc[positive_missing["invoice_pattern"].eq("mixed_missing_present"), "InvoiceStr"].unique()
if len(mixed_invoices) > 0:
    display(
        work.loc[work["InvoiceStr"].isin(mixed_invoices)]
        .groupby("InvoiceStr")
        .agg(
            total_rows=("Customer ID", "size"),
            present_customer=("IsCustomerMissing", lambda s: (~s).sum()),
            missing_customer=("IsCustomerMissing", "sum"),
        )
        .head(20)
    )

# ===== 4. 每张发票的行数分布 =====
display(
    positive_missing.groupby("InvoiceStr")
    .size()
    .describe()
    .to_frame("rows_per_invoice")
)
# 行数分布直方图
display(
    positive_missing.groupby("InvoiceStr")
    .size()
    .value_counts()
    .sort_index()
    .to_frame("invoice_count")
    .head(20)
)

# ===== 5. 时间分布 —— 是某些年份/月份集中吗？ =====
positive_missing = positive_missing.assign(
    year=positive_missing["InvoiceDateParsed"].dt.year,
    month=positive_missing["InvoiceDateParsed"].dt.month,
    quarter=positive_missing["InvoiceDateParsed"].dt.quarter,
)
display(
    positive_missing.groupby(["year", "month"])
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"), invoices=("InvoiceStr", "nunique"))
    .assign(row_pct=lambda x: (x["rows"] / len(positive_missing) * 100).round(2))
)

# ===== 6. 金额区间分布 —— 是小额还是大额？ =====
display(
    positive_missing["LineAmount"]
    .describe([.01, .05, .1, .25, .5, .75, .9, .95, .99])
    .round(2)
    .to_frame("line_amount_distribution")
)
# 按金额区间分桶
bins = [0, 1, 5, 10, 20, 50, 100, 500, 1000, float("inf")]
labels = ["0-1", "1-5", "5-10", "10-20", "20-50", "50-100", "100-500", "500-1k", "1k+"]
display(
    positive_missing.assign(amount_bucket=pd.cut(positive_missing["LineAmount"], bins=bins, labels=labels))
    .groupby("amount_bucket")
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"))
    .assign(row_pct=lambda x: (x["rows"] / len(positive_missing) * 100).round(2))
)

# ===== 7. 单价 vs 数量 —— 有异常点吗？ =====
print("Quantity stats:", positive_missing["Quantity"].describe().round(2).to_dict())
print("Price stats:", positive_missing["Price"].describe([.9, .95, .99]).round(2).to_dict())
# 高单价异常
display(
    positive_missing.loc[positive_missing["Price"].gt(positive_missing["Price"].quantile(0.99))]
    .groupby(["StockCodeStr", "Description"])
    .agg(rows=("Invoice", "size"), avg_price=("Price", "mean"))
    .sort_values("rows", ascending=False)
    .head(15)
    .round(2)
)

# ===== 8. 数量和金额贡献 Top 商品 =====
display(
    positive_missing.groupby(["StockCodeStr", "Description"], dropna=False)
    .agg(
        rows=("Invoice", "size"),
        amount_sum=("LineAmount", "sum"),
        quantity_sum=("Quantity", "sum"),
        invoices=("InvoiceStr", "nunique"),
        avg_price=("Price", "mean"),
    )
    .assign(
        row_pct=lambda x: (x["rows"] / len(positive_missing) * 100).round(2),
        amount_pct=lambda x: (x["amount_sum"] / positive_missing["LineAmount"].sum() * 100).round(2),
    )
    .sort_values("rows", ascending=False)
    .head(20)
    .round(2)
)

,rows,amount_sum,invoices,row_pct,amount_pct
Country,,,,,
United Kingdom,233252,3148203.83,2995,98.78,97.48
EIRE,1606,42800.67,59,0.68,1.33
Hong Kong,358,23894.32,15,0.15,0.74
Unspecified,231,2322.36,8,0.10,0.07
France,128,1687.13,8,0.05,0.05
Switzerland,125,645.95,3,0.05,0.02
Portugal,116,731.35,2,0.05,0.02
United Arab Emirates,84,1894.94,5,0.04,0.06
Bahrain,65,1755.42,6,0.03,0.05


,invoices,pct
customer_id_pattern,,
all_missing,3109,100.0


整单缺失中运单(DOT)占比: 16.1 %


,rows_per_invoice
count,3109.000000
mean,75.947893
std,133.115948
min,1.000000
25%,2.000000
50%,17.000000
75%,109.000000
max,1350.000000


,invoice_count
1,604
2,248
3,134
4,105
5,65
6,61
7,48
8,43
9,41
10,36


rows  amount_sum  invoices  row_pct
year month                                      
2009 12     13203   139031.60       170     5.59
2010 1       8863    95389.44        94     3.75
     2       4915    47342.24        98     2.08
     3       8063   133961.14       157     3.41
     4       6033    86919.80       133     2.56
...           ...         ...       ...      ...
2011 8       7476   113794.48        81     3.17
     9       9233   105751.79        82     3.91
     10      9750   115660.51       111     4.13
     11     18838   347678.95       112     7.98
     12      7807   120599.89        41     3.31

[25 rows x 4 columns]

,line_amount_distribution
count,236122.00
mean,13.68
std,103.48
min,0.06
1%,0.81
5%,0.85
10%,1.63
25%,2.55
50%,5.02
75%,10.79


,rows,amount_sum,row_pct
amount_bucket,,,
0-1,13038,10307.03,5.52
1-5,102411,309667.78,43.37
5-10,56282,399880.28,23.84
10-20,38224,529006.54,16.19
20-50,18808,546692.00,7.97
50-100,4034,274648.32,1.71
100-500,2857,565253.23,1.21
500-1k,328,225737.60,0.14
1k+,140,368346.18,0.06


Quantity stats: {'count': 236122.0, 'mean': 3.02, 'std': 12.12, 'min': 1.0, '25%': 1.0, '50%': 1.0, '75%': 2.0, 'max': 1820.0}
Price stats: {'count': 236122.0, 'mean': 7.05, 'std': 93.58, 'min': 0.06, '90%': 10.79, '95%': 14.95, '99%': 29.17, 'max': 25111.09}


,,rows,avg_price
StockCodeStr,Description,,
DOT,DOTCOM POSTAGE,1289,239.34
M,Manual,123,1531.90
84078A,SET/4 WHITE RETRO STORAGE CUBES,95,64.31
21627,ELEPHANT CARNIVAL POUFFE,44,29.89
22946,WOODEN ADVENT CALENDAR CREAM,38,33.87
21768,WALL MOUNTED VINTAGE ORGANISER,36,29.83
21628,TRIANGULAR POUFFE VINTAGE,32,29.87
21781,MA CAMPAGNE CUTLERY BOX,30,29.89
20784,RED RETRO BAR STOOL,29,35.73


,,rows,amount_sum,quantity_sum,invoices,avg_price,row_pct,amount_pct
StockCodeStr,Description,,,,,,,
DOT,DOTCOM POSTAGE,1420,310751.12,1420,1399,218.84,0.60,9.62
85099B,JUMBO BAG RED RETROSPOT,664,11843.84,3101,633,3.97,0.28,0.37
21931,JUMBO STORAGE BAG SUKI,633,14029.48,3264,615,4.09,0.27,0.43
22423,REGENCY CAKESTAND 3 TIER,633,58076.95,2678,601,23.46,0.27,1.80
47566,PARTY BUNTING,619,45306.82,4771,597,8.98,0.26,1.40
22379,RECYCLING BAG RETROSPOT,573,7031.88,1568,557,4.37,0.24,0.22
22411,JUMBO SHOPPER VINTAGE RED PAISLEY,527,12490.82,3042,512,3.98,0.22,0.39
20719,WOODLAND CHARLOTTE BAG,523,13051.73,5657,511,1.80,0.22,0.40
20712,JUMBO BAG WOODLAND ANIMALS,509,8525.39,2028,500,4.13,0.22,0.26


### 6. 重复行和 Description/StockCode 错配都不能机械处理

精确重复行有 **34,335 行**，但数据没有 invoice line id；同一订单、同一商品、同一时间可能真实出现多行，也可能是重复抽取。不能直接 `drop_duplicates()`，应按下游用途决定：金额报表可做敏感性分析（保留 vs 去重），订单明细则需要原系统行号或额外证据。

`Description` 也不总是商品名。有些 `StockCode` 对应多个描述，例如 `20713` 既是 `JUMBO BAG OWLS`，也出现 `FOUND`、`MISSING`、`WRONGLY CODED`。这说明 `Description` 混入了库存/人工备注，不能直接作为商品维表名称，需要用“最常见正常描述”或外部商品表修正。

In [31]:
duplicate_summary = pd.DataFrame(
    [
        {
            "issue": "exact duplicated rows",
            "rows": int(work.duplicated(subset=df.columns.tolist()).sum()),
            "row_pct": pct(work.duplicated(subset=df.columns.tolist()).sum(), total_rows),
        }
    ]
)
display(duplicate_summary)

display(work.loc[work.duplicated(subset=df.columns.tolist(), keep=False), df.columns.tolist()].head(20))

description_per_stock = (
    work.assign(DescriptionNorm=work["Description"].astype("string").str.strip().str.upper())
    .dropna(subset=["DescriptionNorm"])
    .groupby("StockCodeStr")
    .agg(distinct_descriptions=("DescriptionNorm", "nunique"), rows=("DescriptionNorm", "size"))
    .query("distinct_descriptions > 1")
    .sort_values(["distinct_descriptions", "rows"], ascending=False)
)
display(description_per_stock.head(20))

display(
    work.assign(DescriptionNorm=work["Description"].astype("string").str.strip().str.upper())
    .loc[lambda x: x["StockCodeStr"].eq("20713")]
    .groupby("DescriptionNorm", dropna=False)
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"))
    .sort_values("rows", ascending=False)
)

,issue,rows,row_pct
0,exact duplicated rows,34335,3.22


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


,distinct_descriptions,rows
StockCodeStr,,
20713,8,1380
22423,7,4423
21181,7,1917
23084,7,1057
22734,7,802
47566B,6,801
85175,6,474
21830,6,288
22719,6,260


,rows,amount_sum
DescriptionNorm,,
JUMBO BAG OWLS,1372,40941.11
<NA>,5,0.00
FOUND,2,0.00
MARKED AS 23343,1,0.00
MISSING,1,0.00
WRONGLY CODED 23343,1,0.00
WRONGLY CODED-23343,1,0.00
WRONGLY MARKED 23343,1,0.00
WRONGLY MARKED. 23343 IN BOX,1,0.00


### 7. 缺失 Description 的业务含义

`Description` 缺失 **4,382 行**，但不是普通文本缺失：这些行 **Price 全部为 0**、**Customer ID 全部缺失**，其中约 **61.36%** 是负数量。它们更像库存调整、损耗或系统占位记录，而不是商品销售记录。

处理策略：销售收入/客单价分析应排除；库存质量分析应保留并按 `Quantity` 正负、Invoice、StockCode 单独审计。

In [32]:
description_missing = work["Description"].isna()
description_missing_summary = pd.DataFrame(
    [
        {
            "rows": int(description_missing.sum()),
            "price_zero_rows": int(work.loc[description_missing, "Price"].eq(0).sum()),
            "customer_missing_rows": int(work.loc[description_missing, "Customer ID"].isna().sum()),
            "negative_quantity_rows": int(work.loc[description_missing, "Quantity"].lt(0).sum()),
            "negative_quantity_pct": pct(work.loc[description_missing, "Quantity"].lt(0).sum(), description_missing.sum()),
        }
    ]
)
display(description_missing_summary)
display(work.loc[description_missing, ["Invoice", "StockCode", "Description", "Quantity", "InvoiceDate", "Price", "Customer ID", "Country"]].head(20))

,rows,price_zero_rows,customer_missing_rows,negative_quantity_rows,negative_quantity_pct
0,4382,4382,4382,2689,61.36


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom
6378,489882,35751C,NaN,12,2009-12-02 16:22:00,0.0,NaN,United Kingdom
6555,489898,79323G,NaN,954,2009-12-03 09:40:00,0.0,NaN,United Kingdom
6576,489901,21098,NaN,-200,2009-12-03 09:47:00,0.0,NaN,United Kingdom
6581,489903,21166,NaN,48,2009-12-03 09:57:00,0.0,NaN,United Kingdom


### 8. 时间范围影响月度趋势可比性

数据范围是 **2009-12-01 07:45:00 到 2011-12-09 12:50:00**。因此 2009 年 12 月从月初开始但未必覆盖业务完整历史；**2011 年 12 月只到 12 月 9 日**，不是完整月份。后续如果做月度趋势、同比或季节性分析，不能把 2011-12 直接和其他完整月份比较。

In [33]:
date_range_summary = pd.DataFrame(
    [
        {
            "min_invoice_date": work["InvoiceDateParsed"].min(),
            "max_invoice_date": work["InvoiceDateParsed"].max(),
            "last_month": work["InvoiceDateParsed"].max().strftime("%Y-%m"),
            "last_month_last_day_in_data": work["InvoiceDateParsed"].max().day,
        }
    ]
)
display(date_range_summary)

display(
    work.assign(month=work["InvoiceDateParsed"].dt.to_period("M"))
    .groupby("month")
    .agg(rows=("Invoice", "size"), amount_sum=("LineAmount", "sum"))
    .tail(6)
)

,min_invoice_date,max_invoice_date,last_month,last_month_last_day_in_data
0,2009-12-01 07:45:00,2011-12-09 12:50:00,2011-12,9


,rows,amount_sum
month,,
2011-07,39518,681300.111
2011-08,35284,682680.510
2011-09,50226,1019687.622
2011-10,60742,1070704.670
2011-11,84711,1461756.250
2011-12,25526,433686.010


## Q1.1 小结：非显而易见的数据质量问题与处理策略

1. **Invoice 前缀有交易语义**：`C` 是取消/退货，`A` 是 bad debt adjustment。不要把所有行混在一个销售总额里；先定义 gross product sales、net product sales 和财务总额口径。
2. **特殊 StockCode 会污染商品分析**：邮费、折扣、manual adjustment、平台费用/佣金、礼品券、样品/测试/内部代码应单独分组。商品销量、热销 SKU、商品价格分析应排除这些非产品行；财务分析可以保留但要分项列示。
3. **极端 Quantity 有业务解释**：`80995` 和 `-80995` 是同一客户同一商品的销售与取消，间隔 12 分钟。它们不适合直接删除；做净销售时可互相抵消，做异常监控时应作为大额取消案例保留。
4. **负价格、零价格、负数量不是同一类问题**：负价格基本是 bad debt；零价格多是 lost/damaged/damages/wet/short/check/缺失描述；数字 Invoice 的负数量且零价格多为商品损坏或库存损耗。收入分析应排除这类零价损耗记录，库存质量分析应保留。
5. **Customer ID 缺失不是随机缺失**：缺失行占 22.77%，但金额占 13.68%，98.77% 来自 UK，平均行金额更低且高频出现 `DOTCOM POSTAGE`。客户分析要剔除或标记 anonymous/system channel；销售收入可保留但需单独披露。
6. **精确重复行不能无脑删除**：有 34,335 行精确重复，但缺少 invoice line id，无法判断是真重复还是同一订单同一商品被真实拆成多行。建议金额报表同时给出保留/去重敏感性，最终去重需业务主键支持。
7. **Description 混入库存/人工备注**：同一 StockCode 可能对应商品名和 `FOUND/MISSING/WRONGLY CODED/DAMAGED` 等备注。商品维表应优先使用 StockCode，并用最常见正常描述或外部商品表标准化名称。
8. **Description 缺失有明确模式**：缺失描述行全部零价格、全部缺 Customer ID，61.36% 为负数量。它们应从销售分析中排除，但可进入库存损耗/数据录入质量分析。
9. **2011-12 不是完整月份**：月度趋势、同比和季节性分析要排除或截断不完整月份，避免把 2011 年 12 月误判为销售下降。

**推荐清洗分层**：

- **商品销售 gross 口径**：保留数字 Invoice、产品型 StockCode、`Quantity > 0`、`Price > 0`；排除 `A/C`、非产品 StockCode、零价商品损坏/库存损耗、缺失 Description 的系统调整。
- **商品销售 net 口径**：在 gross 基础上纳入 `C` 取消/退货作为抵减，但仍排除 bad debt、平台费用、邮费、折扣、manual adjustment 等非商品行。
- **客户/RFM 口径**：仅使用有 Customer ID 的商品销售/退货记录；缺失 Customer ID 可作为 anonymous/system channel 单独统计，不混入客户生命周期指标。
- **库存/运营质量口径**：保留 `IsDamageOrInventoryLoss`、零价格、负数量、缺失 Description、FOUND/MISSING/DAMAGED/WRONGLY CODED 等记录，用于识别商品损坏、库存损耗、盘点和人工录入问题。
- **财务总额口径**：保留所有金额行，但必须把 product sales、returns、postage/carriage、discount、fees/commission、bad debt 分组展示。